### PR TITLE
Focus change plugin added workspace ignore flag and changed some logic in regards to cross-output functionality

### PR DESCRIPTION
**Pull request by Kim-Dewelski**
changes how outputs are selected when using cross-output and added a setting to optionally ignore workspace boundaries (views can only be focused if it's in the current workspace)

Overall I consider this code an improvement and fixes some annoyances I had with the plugin where it would not respect workspace boundaries and it was impossible to change via an option.

### CODE REVIEW

File: src/focus-change.cpp [hunk 1]
```diff
@@ -155,19 +152,124 @@ class wayfire_focus_change_t : public wf::plugin_interface_t
...
+                        new_focus   = view.get();
+                    }
                 }
             }
         }
 
         if (new_focus != nullptr)
         {
-            wf::get_core().seat->focus_output(new_focus->get_output());
+            auto op = new_focus->get_output();
```
Comment (soreau) [inline src/focus-change.cpp:269]: As a 'gotcha', view->get_output(); can return NULL, so we need to handle this case.

File: src/focus-change.cpp [hunk 2]
```diff
@@ -155,19 +152,124 @@ class wayfire_focus_change_t : public wf::plugin_interface_t
...
+            const auto ws = op->wset()->get_view_main_workspace(new_focus);
+            op->wset()->set_workspace(ws);
```
Comment (soreau) [inline src/focus-change.cpp:271]: If I remember correctly, there is wset()->request_workspace() which should get you the workspace transition animation, at least.

File: src/focus-change.cpp [hunk 3]
```diff
@@ -124,11 +121,11 @@ class wayfire_focus_change_t : public wf::plugin_interface_t
...
             const int32_t bias_left  = grace_left.value();
 
-            const bool w_cond = cx >= scan_w_l && cx <= scan_w_h;
-            const bool h_cond = cy >= scan_h_l && cy <= scan_h_h;
+            const bool w_cond = cx + cxr >= scan_w_l && cx - cxr <= scan_w_h;
+            const bool h_cond = cy + cyr >= scan_h_l && cy - cyr <= scan_h_h;
 
             bool contains    = false;
-            int32_t distance = 0;
+            int32_t distance = INT32_MAX;
```
Comment (soreau) [inline src/focus-change.cpp:128]: What is the reasoning behind switching to INT32_MAX and using it in other places? Would initializing to 0 make things not work correctly?

**Review by soreau — CHANGES_REQUESTED** [review]
Looks good, just a few things I have questions about.

File: src/focus-change.cpp (hunk 1 above)
Comment (Kim-Dewelski) [reply src/focus-change.cpp:269]: Nice I will fix that

File: src/focus-change.cpp (hunk 2 above)
Comment (Kim-Dewelski) [reply src/focus-change.cpp:271]: Nice, I will use this instead!

File: src/focus-change.cpp (hunk 3 above)
Comment (Kim-Dewelski) [reply src/focus-change.cpp:128]: there was no difference per say, it was just to make it consistent with choices I made elsewhere.

**Comment by Kim-Dewelski** [comment]
Actally, I use get_output() elsewhere, I will have to add some checks.

**Comment by Kim-Dewelski** [comment]
hopefully this is a bit better now?

**Comment by Kim-Dewelski** [comment]
@soreau any chance to merge this asap?

**Review by soreau — APPROVED** [review]

**Pull request merged**